test/Directory.test.ts

```diff
@@ -0,0 +1,640 @@
+    await startProtocol();
+    await stakingOrchestator.syloStakeAdded(nodeOne, ethers.ZeroAddress, 1000);
+    await directory.connect(nodeOne).joinNextDirectory();
+    await expect(
+      directory.connect(nodeOne).joinNextDirectory(),
+    ).to.be.revertedWithCustomError(directory, 'StakeeAlreadyJoinedDirectory');
+  });
+
+  it('should be able to scan after joining directory', async () => {
+    const { setTimeSinceStart } = await startProtocol();
+    await stakingOrchestator.syloStakeAdded(nodeOne, ethers.ZeroAddress, 1000);
+    await directory.connect(nodeOne).joinNextDirectory();
+    await setTimeSinceStart(110);
+    const address = await directory.scan(0);
+    assert.equal(address, await nodeOne.getAddress());
+  });
+
+  it('should be able to scan with period id after joining directory', async () => {
+    await startProtocol();
+    await stakingOrchestator.syloStakeAdded(nodeOne, ethers.ZeroAddress, 1000);
+    await directory.connect(nodeOne).joinNextDirectory();
+    const address = await directory.scanWithTime(0, 1, 1);
+    assert.equal(address, await nodeOne.getAddress());
+  });
+
+  it('should be able to scan empty directory', async () => {
+    await startProtocol();
+    const address = await directory.scan(0);
+    assert.equal(address.toString(), ethers.ZeroAddress);
+  });
+
+  it('should be able to scan for different staking periods', async () => {
+    let address: string;
+    const { setTimeSinceStart } = await startProtocol();
+    await stakingOrchestator.syloStakeAdded(nodeOne, ethers.ZeroAddress, 1000);
+    await directory.connect(nodeOne).joinNextDirectory();
+
+    await setTimeSinceStart(150);
+
+    address = await directory.scan(0);
+    assert.equal(address, await nodeOne.getAddress());
+
+    await setTimeSinceStart(500);
+
+    address = await directory.scan(0);
+    assert.equal(address, ethers.ZeroAddress);
+  });
+
+  it('node joins next reward cycle', async () => {
+    let address: string;
+    const { setTimeSinceStart } = await startProtocol();
+    await stakingOrchestator.syloStakeAdded(nodeOne, ethers.ZeroAddress, 1000);
+    await directory.connect(nodeOne).joinNextDirectory();
+
+    await setTimeSinceStart(150);
+
+    address = await directory.scan(0);
+    assert.equal(address, await nodeOne.getAddress());
+
+    await setTimeSinceStart(850);
+
+    await directory.connect(nodeOne).joinNextDirectory();
+
+    await setTimeSinceStart(950);
+
+    address = await directory.scan(0);
+    assert.equal(address, await nodeOne.getAddress());
+
+    await setTimeSinceStart(1050);
+
+    address = await directory.scan(0);
+    assert.equal(address, ethers.ZeroAddress);
+
+    await setTimeSinceStart(1950);
+    await directory.connect(nodeOne).joinNextDirectory();
+
+    await setTimeSinceStart(2050);
+
+    address = await directory.scan(0);
+    assert.equal(address, await nodeOne.getAddress());
+  });
+
+  it('should be able to scan for same staking period', async () => {
```

Comment:
```suggestion
  it('should be able to scan for previous periods', async () => {
```

This seems like what the test is actually checking for

Reply:
But I would suggest using multiple nodes, and make sure each period has a different set of nodes join (so the directories are different for each period). Reason being that if you only had 1 node joining, and your `scanWithTime` implementation was wrong in the sense that it always scanned the current period instead of the specified time, then this test would pass.